### PR TITLE
Add support for CANFD and Disabling Tail Array Optimisation

### PR DESCRIPTION
**Pull request by bugobliterator**

### CODE REVIEW

**Comment by tridge** [comment]
@bugobliterator nice work!

**Comment by tridge** [comment]
@bugobliterator what testing has been done on this when CANFD is not enabled?

File: libuavcan/include/uavcan/driver/can.hpp [hunk 1]
```diff
@@ -42,15 +47,42 @@ struct UAVCAN_EXPORT CanFrame
...
-    CanFrame(uint32_t can_id, const uint8_t* can_data, uint8_t data_len) :
+    CanFrame(uint32_t can_id, const uint8_t* can_data, uint8_t data_len, bool canfd_frame = false) :
         id(can_id),
-        dlc((data_len > MaxDataLen) ? MaxDataLen : data_len)
+        canfd(canfd_frame)
     {
         UAVCAN_ASSERT(can_data != UAVCAN_NULLPTR);
         UAVCAN_ASSERT(data_len == dlc);
-        (void)copy(can_data, can_data + dlc, this->data);
+        (void)copy(can_data, can_data + data_len, this->data);
```
Comment (tridge) [inline libuavcan/include/uavcan/driver/can.hpp:56]: this logic needs to be moved to the cpp file

File: libuavcan/include/uavcan/driver/can.hpp [hunk 2]
```diff
@@ -61,6 +93,58 @@ struct UAVCAN_EXPORT CanFrame
     bool isRemoteTransmissionRequest() const { return id & FlagRTR; }
     bool isErrorFrame()                const { return id & FlagERR; }
 
+    static uint8_t dlcToDataLength(uint8_t dlc) {
```
Comment (tridge) [inline libuavcan/include/uavcan/driver/can.hpp:96]: should be in cpp file

File: libuavcan/include/uavcan/driver/can.hpp [hunk 3]
```diff
@@ -61,6 +93,58 @@ struct UAVCAN_EXPORT CanFrame
...
+            return 24;
+        } else if (dlc == 13) {
+            return 32;
+        } else if (dlc == 14) {
+            return 48;
+        }
+        return 64;
+    }
+
+    static uint8_t dataLengthToDlc(uint8_t data_length) {
```
Comment (tridge) [inline libuavcan/include/uavcan/driver/can.hpp:119]: in cpp file

File: libuavcan/include/uavcan/driver/can.hpp [hunk 4]
```diff
@@ -61,6 +93,58 @@ struct UAVCAN_EXPORT CanFrame
...
+            return 12;
+        } else if (data_length <= 32) {
+            return 13;
+        } else if (data_length <= 48) {
+            return 14;
+        }
+        return 15;
+    }
+
+    static uint8_t getNumPaddingBytes(uint16_t payload_len) {
```
Comment (tridge) [inline libuavcan/include/uavcan/driver/can.hpp:138]: in cpp file

**Review by tridge — CHANGES_REQUESTED** [review]
i think the inlining is one of the reason the PR adds so much flash

File: libuavcan/include/uavcan/node/abstract_node.hpp [hunk 5]
```diff
@@ -73,6 +77,8 @@ class UAVCAN_EXPORT INode
      */
     int spin(MonotonicDuration duration)
     {
+        getScheduler().getDispatcher().setTaoDisabled(tao_disabled_);
```
Comment (tridge) [inline libuavcan/include/uavcan/node/abstract_node.hpp:80]: maybe: getScheduler().getDispatcher().set_options(tao_disabled_, tao_disabled_);

File: libuavcan/include/uavcan/node/generic_publisher.hpp [hunk 6]
```diff
@@ -165,7 +168,9 @@ int GenericPublisher<DataSpec, DataStruct>::doEncode(const DataStruct& message,
 {
     BitStream bitstream(buffer);
     ScalarCodec codec(bitstream);
-    const int encode_res = DataStruct::encode(message, codec);
+    // if doing canfd transfer tail array optimisation is disabled
+    TailArrayOptimizationMode tao_mode = (!isStdOnly() && (getNode().isTaoDisabled() || getNode().isCanFdEnabled())) ? TailArrayOptDisabled:TailArrayOptEnabled;
```
Comment (tridge) [inline libuavcan/include/uavcan/node/generic_publisher.hpp:172]: add a get_tao_mode() call

**Review by tridge — CHANGES_REQUESTED** [review]

File: libuavcan/include/uavcan/driver/can.hpp (hunk 2 above)
Comment (bugobliterator) [reply libuavcan/include/uavcan/driver/can.hpp:96]: doing that increases the flash cost

File: libuavcan/include/uavcan/driver/can.hpp (hunk 3 above)
Comment (bugobliterator) [reply libuavcan/include/uavcan/driver/can.hpp:119]: doing that increases the flash cost

File: libuavcan/include/uavcan/driver/can.hpp (hunk 4 above)
Comment (bugobliterator) [reply libuavcan/include/uavcan/driver/can.hpp:138]: doing that increases the flash cost

File: libuavcan/include/uavcan/driver/can.hpp (hunk 1 above)
Comment (bugobliterator) [reply libuavcan/include/uavcan/driver/can.hpp:56]: doing that costs extra flash

File: libuavcan/include/uavcan/node/generic_publisher.hpp (hunk 6 above)
Comment (bugobliterator) [reply libuavcan/include/uavcan/node/generic_publisher.hpp:172]: this is the only place this is used, and information is gathered from different scopes.

File: libuavcan/include/uavcan/node/abstract_node.hpp (hunk 5 above)
Comment (bugobliterator) [reply libuavcan/include/uavcan/node/abstract_node.hpp]: done

File: libuavcan/src/transport/uc_frame.cpp [hunk 7]
```diff
@@ -159,11 +161,20 @@ bool Frame::compile(CanFrame& out_can_frame) const
...
     (void)copy(payload_, payload_ + payload_len_, out_can_frame.data);
 
-    out_can_frame.data[out_can_frame.dlc] = tail;
-    out_can_frame.dlc++;
+    if (payload_len_ < 8) {
+        out_can_frame.data[payload_len_] = tail;
+        out_can_frame.dlc++;
+    } else if (payload_len_ == CanFrame::dlcToDataLength(out_can_frame.dlc)){
+        out_can_frame.dlc++;
+        out_can_frame.data[CanFrame::dlcToDataLength(out_can_frame.dlc) - 1] = tail;
```
Comment (tridge) [inline libuavcan/src/transport/uc_frame.cpp:172]: do we need to memset here to zero the pad bytes? Are all frames guaranteed to start with zeros in the data field? I'm concerned we could leak data

**Review by tridge — CHANGES_REQUESTED** [review]

**Review by tridge — APPROVED** [review]

**Pull request merged**